mappers/modbus/device/device.go

```diff
@@ -0,0 +1,251 @@
+			Parity:       protocolConfig.Com.Parity,
+			RS485Enabled: isRS485Enabled(protocolConfig.CustomizedValues),
+			Timeout:      5 * time.Second}
+		*client, _ = driver.NewClient(modbusRtu)
+	} else if protocolConfig.Tcp.Ip != "" {
+		modbusTcp := driver.ModbusTcp{
+			SlaveId:  byte(slaveID),
+			DeviceIp: protocolConfig.Tcp.Ip,
+			TcpPort:  strconv.FormatInt(protocolConfig.Tcp.Port, 10),
+			Timeout:  5 * time.Second}
+		*client, _ = driver.NewClient(modbusTcp)
+	} else {
+		return errors.New("No protocol found")
+	}
+	return nil
+}
+
+// initTwin initialize the timer to get twin value.
+func initTwin(dev *globals.ModbusDev) {
+	for i := 0; i < len(dev.Instance.Twins); i++ {
+		var visitorConfig configmap.ModbusVisitorConfig
+		if err := json.Unmarshal([]byte(dev.Instance.Twins[i].PVisitor.VisitorConfig), &visitorConfig); err != nil {
+			klog.Error(err)
+			continue
+		}
+		value, err := mappercommon.Convert(dev.Instance.Twins[i].PVisitor.PProperty.DataType, dev.Instance.Twins[i].Desired.Value)
+		if err != nil {
+			klog.Error("Convert error. Type: %s, value: %s ", dev.Instance.Twins[i].PVisitor.PProperty.DataType, dev.Instance.Twins[i].Desired.Value)
+			continue
+		}
+		valueInt, _ := value.(int64)
+		_, err = dev.ModbusClient.Set(visitorConfig.Register, visitorConfig.Offset, uint16(valueInt))
+		if err != nil {
+			klog.Error(err, visitorConfig)
+			continue
+		}
+
+		twinData := TwinData{Client: dev.ModbusClient,
+			Name:         dev.Instance.Twins[i].PropertyName,
+			Type:         dev.Instance.Twins[i].Desired.Metadatas.Type,
+			RegisterType: visitorConfig.Register,
+			Address:      visitorConfig.Offset,
+			Quantity:     uint16(visitorConfig.Limit),
+			Topic:        fmt.Sprintf(mappercommon.TopicTwinUpdate, dev.Instance.ID)}
+		collectCycle := time.Duration(dev.Instance.Twins[i].PVisitor.CollectCycle)
+		// If the collect cycle is not set, set it to 1 second.
+		if collectCycle == 0 {
+			collectCycle = 1 * time.Second
+		}
+		timer := mappercommon.Timer{twinData.Run, collectCycle, 0}
+		go timer.Start()
+		wg.Add(1)
+	}
+}
+
+// initData initialize the timer to get data.
+func initData(dev *globals.ModbusDev) {
+	for i := 0; i < len(dev.Instance.Datas.Properties); i++ {
+		var visitorConfig configmap.ModbusVisitorConfig
+		if err := json.Unmarshal([]byte(dev.Instance.Datas.Properties[i].PVisitor.VisitorConfig), &visitorConfig); err != nil {
+			klog.Error("Unmarshal visitor config failed")
+		}
+		twinData := TwinData{Client: dev.ModbusClient,
+			Name:         dev.Instance.Datas.Properties[i].PropertyName,
+			Type:         dev.Instance.Datas.Properties[i].Metadatas.Type,
+			RegisterType: visitorConfig.Register,
+			Address:      visitorConfig.Offset,
+			Quantity:     uint16(visitorConfig.Limit),
+			Topic:        dev.Instance.Datas.Topic}
+		timer := mappercommon.Timer{twinData.Run, time.Duration(dev.Instance.Twins[i].PVisitor.CollectCycle), 0}
+		go timer.Start()
+		wg.Add(1)
```

Comment:
it's better to close goroutine when it has finshed it's work .So just add
```
defer wg.Done()
```

Reply:
Done.